mate-panel/menu.c

```diff
@@ -311,36 +305,36 @@ add_drawers_from_dir (MateMenuTreeDirectory *directory,
 	if (!attached_toplevel_id)
 		return;
 
-	items = matemenu_tree_directory_get_contents (directory);
-	for (l = items; l; l = l->next) {
-		MateMenuTreeItem *item = l->data;
-
-		switch (matemenu_tree_item_get_type (item)) {
-		case MATEMENU_TREE_ITEM_ENTRY:
-			panel_launcher_create_with_id (attached_toplevel_id,
-						       G_MAXINT/2,
-						       matemenu_tree_entry_get_desktop_file_path (MATEMENU_TREE_ENTRY (item)));
-			break;
-
-		case MATEMENU_TREE_ITEM_DIRECTORY:
-			add_drawers_from_dir (MATEMENU_TREE_DIRECTORY (item),
-					      G_MAXINT/2,
-					      attached_toplevel_id);
-			break;
-
-		case MATEMENU_TREE_ITEM_ALIAS:
-			add_drawers_from_alias (MATEMENU_TREE_ALIAS (item), attached_toplevel_id);
-			break;
-
-		default:
-			break;
+	iter = matemenu_tree_directory_iter (directory);
+	while ((type = matemenu_tree_iter_next (iter)) != MATEMENU_TREE_ITEM_INVALID) {
+		gpointer item;
+		switch (type) {
+			case MATEMENU_TREE_ITEM_ENTRY:
+				item = matemenu_tree_iter_get_entry (iter);
+				panel_launcher_create_with_id (attached_toplevel_id,
+						G_MAXINT/2,
+						matemenu_tree_entry_get_desktop_file_path (item));
+				matemenu_tree_item_unref (item);
+				break;
+
+			case MATEMENU_TREE_ITEM_DIRECTORY:
+				item = matemenu_tree_iter_get_directory (iter);
+				add_drawers_from_dir (item,
+						G_MAXINT/2,
+						attached_toplevel_id);
+				matemenu_tree_item_unref (item);
+				break;
+
+			case MATEMENU_TREE_ITEM_ALIAS:
+				item = matemenu_tree_iter_get_alias (iter);
+				add_drawers_from_alias (item, attached_toplevel_id);
+				matemenu_tree_item_unref (item);
+				break;
+			default:
+				g_assert_not_reached ();
+				break;
 		}
-
-		matemenu_tree_item_unref (item);
 	}
-
-	g_slist_free (items);
```

Comment:
Iterator created by `matemenu_tree_directory_iter` should be freed with `matemenu_tree_iter_unref`.